pkg/datapath/tables/device.go

```diff
@@ -0,0 +1,115 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package tables
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+
+	"github.com/hashicorp/go-memdb"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/exp/slices"
+
+	"github.com/cilium/cilium/pkg/statedb"
+)
+
+const (
+	DeviceNameIndex statedb.Index = "Name"
+)
+
+var deviceTableSchema = &memdb.TableSchema{
+	Name: "devices",
+	Indexes: map[string]*memdb.IndexSchema{
+		string(statedb.IDIndex): {
+			Name:         string(statedb.IDIndex),
+			AllowMissing: false,
+			Unique:       true,
+			Indexer:      &memdb.IntFieldIndex{Field: "Index"},
+		},
+		string(DeviceNameIndex): {
+			Name:         string(DeviceNameIndex),
+			AllowMissing: false,
+			Unique:       true,
+			Indexer:      &memdb.StringFieldIndex{Field: "Name"},
+		},
+	},
+}
+
+// Device is a local network device along with addresses associated with it.
+type Device struct {
+	Index  int             // Interface index (primary key)
+	Name   string          // Interface name (e.g. eth0)
+	Viable bool            // If true this device can be used by Cilium
+	Addrs  []DeviceAddress // Addresses assigned to the device
+	Link   netlink.Link    // The underlying link device
+}
+
+func (d *Device) String() string {
+	return fmt.Sprintf("Device{Index:%d, Name:%s, len(Addrs):%d}", d.Index, d.Name, len(d.Addrs))
+}
+
+func (d *Device) DeepCopy() *Device {
+	copy := *d
+	copy.Addrs = slices.Clone(d.Addrs)
+	return &copy
+}
+
+func (d *Device) HasIP(ip net.IP) bool {
+	for _, addr := range d.Addrs {
+		if addr.AsIP().Equal(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+type DeviceAddress struct {
+	Addr  netip.Addr
+	Scope int // Routing table scope
+	Flags int // Address flags
+}
+
+func (d *DeviceAddress) AsIP() net.IP {
+	return d.Addr.AsSlice()
+}
+
+// DeviceByIndex constructs a query to find a device by its
+// interface index.
+func DeviceByIndex(index int) statedb.Query {
+	return statedb.Query{Index: statedb.IDIndex, Args: []any{index}}
+}
+
+// DeviceByName constructs a query to find a device by its
+// name.
+func DeviceByName(name string) statedb.Query {
+	return statedb.Query{Index: DeviceNameIndex, Args: []any{name}}
+}
+
+// ViableDevices returns all current viable network devices.
+// The invalidated channel is closed when devices have changed and
+// should be requeried with a new transaction.
+func ViableDevices(r statedb.TableReader[*Device]) (devs []*Device, invalidated <-chan struct{}) {
+	iter, err := r.Get(statedb.Query{Index: DeviceNameIndex})
+	if err != nil {
+		// table schema is malformed?
+		panic(err)
```

Comment:
Yeah, that's the unfortunate effect of the very dynamic indexer functions in go-memdb: they reflect on the arguments and potentially accept many different forms. The failure in Get() is only if something unexpected is passed to the indexers `FromArgs` or an unknown index is used. 

My thinking here was that since this file both defines what the indexes are and this function, it is fine for this function to make the assumption that it does construct a correct query and thus panicing is fine. Any bug here would be triggered in tests anyway.

For the general case it'd be great to find a good way to construct queries for statedb tables in a safe way that cannot cause errors, but I'm not quite sure how to do that in an ergonomic way without getting overly fancy with generic types.